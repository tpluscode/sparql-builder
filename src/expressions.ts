export { IN } from './lib/expressions/in'
export { VALUES } from './lib/expressions/values'
