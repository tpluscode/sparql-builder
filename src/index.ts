export { sparql } from '@tpluscode/rdf-string'
export { DELETE } from './lib/DeleteBuilder.js'
export type { DeleteInsertQuery as DeleteInsert, DeleteData } from './lib/DeleteBuilder.js'
export { INSERT } from './lib/InsertBuilder.js'
export type { InsertQuery as Insert, InsertData } from './lib/InsertBuilder.js'
export { ASK } from './lib/AskBuilder.js'
export type { AskQuery as Ask } from './lib/AskBuilder.js'
export { DESCRIBE } from './lib/DescribeBuilder.js'
export type { DescribeQuery as Describe } from './lib/DescribeBuilder.js'
export { CONSTRUCT } from './lib/ConstructBuilder.js'
export type { ConstructQuery as Construct } from './lib/ConstructBuilder.js'
export { SELECT } from './lib/SelectBuilder.js'
export type { SelectQuery as Select } from './lib/SelectBuilder.js'
export { WITH } from './lib/WithBuilder.js'
export { prefixes } from '@tpluscode/rdf-string'
export type { SparqlTemplateResult } from '@tpluscode/rdf-string'
export type {
  SparqlAskExecutable, SparqlGraphQueryExecutable, SparqlQueryExecutable, SparqlUpdateExecutable,
} from './lib/index.js'
