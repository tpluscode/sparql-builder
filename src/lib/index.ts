import { AskQuery, ConstructQuery, QueryOptions, SelectQuery, UpdateQuery } from 'sparql-http-client'
import { SparqlTemplateResult } from '@tpluscode/rdf-string'
import prologue, { PrologueBuilder } from './partials/prologue'

interface SparqlBuildOptions {
  base?: string
}

export type SparqlExecuteOptions = QueryOptions & SparqlBuildOptions

export interface SparqlQuery extends PrologueBuilder {
  build(options?: SparqlBuildOptions): string
  _getTemplateResult(): SparqlTemplateResult
}

export interface SparqlQueryExecutable {
  execute<TQuery extends SelectQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['select']>
}

export interface SparqlGraphQueryExecutable {
  execute<TQuery extends ConstructQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['construct']>
}

export interface SparqlUpdateExecutable {
  execute<TQuery extends UpdateQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['update']>
}

export interface SparqlAskExecutable {
  execute<TQuery extends AskQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['ask']>
}

type TBuilder = Pick<SparqlQuery, 'build'> & Pick<SparqlTemplateResult, '_toPartialString'>

// eslint-disable-next-line no-unused-vars
export default function Builder<T extends SparqlQuery>(): TBuilder & T {
  return {
    ...prologue(),
    build(this: SparqlQuery, { base }: SparqlBuildOptions = {}): string {
      const queryResult = this._getTemplateResult().toString({
        base,
      })

      if (this.prologueResult) {
        return `${this.prologueResult}\n\n${queryResult}`
      }

      return queryResult
    },
    _toPartialString(this: SparqlQuery, options: any) {
      return this._getTemplateResult()._toPartialString(options)
    },
  } as any
}
