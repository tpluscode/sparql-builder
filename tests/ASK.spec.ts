import namespace from '@rdfjs/namespace'
import { namedNode } from '@rdfjs/data-model'
import { ASK } from '../src'
import { sparqlClient } from './_mocks'

describe('ASK', () => {
  it('creates expected query', () => {
    // given
    const expected = `ASK {
      ?s ?p ?o .
      ?x ?y ?z .
    }`

    // when
    const query = ASK`?s ?p ?o . ?x ?y ?z .`.build()

    // then
    expect(query).toMatchQuery(expected)
  })

  it('executes as ask', async () => {
    // given
    const client = sparqlClient()

    // when
    await ASK``.execute(client)

    // then
    expect(client.ask).toHaveBeenCalled()
  })

  it('can have additional prologue', () => {
    // given
    const base = namedNode('http://foo.bar/baz')

    // when
    const query = ASK`?s ?p ?o .`
      .prologue`#pragma join.hash off`
      .prologue`BASE ${base}`
      .build()

    // then
    expect(query).toMatchSnapshot()
  })

  it('supports LIMIT/OFFSET', () => {
    // given
    const expected = `ASK {
      ?s ?p ?o .
    } LIMIT 10 OFFSET 20`

    // when
    const query = ASK`?s ?p ?o .`.LIMIT(10).OFFSET(20).build()

    // then
    expect(query).toMatchQuery(expected)
  })

  it('supports FROM (NAMED)', () => {
    // given
    const expected = `ASK 
      FROM <http://example.com/foo>
      FROM NAMED <http://example.com/bar>
      {
        ?s ?p ?o .
      }`

    // when
    const query = ASK`?s ?p ?o .`
      .FROM(namedNode('http://example.com/foo'))
      .FROM().NAMED(namedNode('http://example.com/bar'))
      .build()

    // then
    expect(query).toMatchQuery(expected)
  })

  it('can be constructed with a base', () => {
    // given
    const ns = namespace('http://example.com/')
    const expected = `BASE <http://example.com/>

    ASK {
      <person> a <Person>
    }`

    // when
    const query = ASK`${ns.person} a ${ns.Person} .`.build({
      base: 'http://example.com/',
    })

    // then
    expect(query).toMatchQuery(expected)
  })
})
