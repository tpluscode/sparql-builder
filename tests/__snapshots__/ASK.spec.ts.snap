// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ASK can have additional prologue 1`] = `
"#pragma join.hash off
BASE <http://foo.bar/baz>

ASK  { ?s ?p ?o . }
"
`;
